extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/TransferObject.java

```diff
@@ -0,0 +1,122 @@
+                final ObjectCodec<BigDecimal> codec = new BigDecimalParquetBytesCodec(
+                        precisionAndScale.precision, precisionAndScale.scale, -1);
+                return new CodecTransfer<>(bigDecimalColumnSource, codec, maxValuesPerPage,
+                        instructions.getTargetPageSize());
+            } else if (BigInteger.class.equals(columnType)) {
+                return new CodecTransfer<>(columnSource, new BigIntegerParquetBytesCodec(-1),
+                        maxValuesPerPage, instructions.getTargetPageSize());
+            }
+        }
+
+        final ObjectCodec<? super DATA_TYPE> codec = CodecLookup.lookup(columnDefinition, instructions);
+        return new CodecTransfer<>(columnSource, codec, maxValuesPerPage, instructions.getTargetPageSize());
+    }
+
+    /**
+     * Fetch all data corresponding to the provided row sequence.
+     */
+    void fetchData(@NotNull RowSequence rs);
+
+    /**
+     * Transfer all the fetched data into an internal buffer, which can then be accessed using
+     * {@link TransferObject#getBuffer()}. This method should only be called after
+     * {@link TransferObject#fetchData(RowSequence)}}. This method should be used when writing unpaginated data, and
+     * should not be interleaved with calls to {@link TransferObject#transferOnePageToBuffer()}. Note that this
+     * method can lead to out-of-memory error for variable-width types (e.g. strings) if the fetched data is too big
+     * to fit in the available heap.
+     *
+     * @return The number of fetched data entries copied into the buffer.
+     */
+    int transferAllToBuffer();
+
+    /**
+     * Transfer one page size worth of fetched data into an internal buffer, which can then be accessed using
+     * {@link TransferObject#getBuffer()}. The target page size is passed in the constructor. The method should only
+     * be called after {@link TransferObject#fetchData(RowSequence)}}. This method should be used when writing
+     * paginated data, and should not be interleaved with calls to {@link TransferObject#transferAllToBuffer()}.
+     *
+     * @return The number of fetched data entries copied into the buffer. This can be different from the total
+     * number of entries fetched in case of variable-width types (e.g. strings) when used with additional
+     * page size limits while copying.
+     */
+    int transferOnePageToBuffer();
+
+    /**
+     * Check if there is any fetched data which can be copied into buffer
+     */
+    boolean hasMoreDataToBuffer();
+
+    /**
+     * Get the buffer suitable for writing to a Parquet file
+     *
+     * @return the buffer
+     */
+    B getBuffer();
+
+    /**
+     * After a sequence of calls to {@link #fetchData(RowSequence)} this may be called to append statistic information
+     * into the Parquet footer for readers.
+     */
+    <T extends Comparable<T>> void updateStatistics(@NotNull Statistics<T> stats);
```

Comment:
This is done at a lower level, internal to the `io.deephaven.parquet.base.BulkWriter` implementations. Replicating statistics generation at this level adds negative value unless we clean up the lower level as well, but the lower level is doing a better job (tracking null count, for example).